File: 2024/February/2-16-24/LeastUniqueInts_1481.java
import java.util.HashMap;
import java.util.Collections;
import java.util.ArrayList;

public class LeastUniqueInts_1481 {
	public static void main(String[] args) {
		int[] nums = {4, 3, 1, 1, 3, 3, 2};
		System.out.println(new Solution().findLeastNumOfUniqueInts(nums, 3));
	}
}

class Solution {
	public int findLeastNumOfUniqueInts(int[] arr, int k) {
		HashMap<Integer, Integer> numCounts = new HashMap<Integer, Integer>();	// Stores the mapping of int values to their count
		ArrayList<Integer> counts;	// ArrayList built from the values in the numCounts map
		int uniqueCount;	// Number of unique values in arr
	
		/* Get count of values in arr */
		for(int i : arr) {
			numCounts.put(i, numCounts.getOrDefault(i, 0) + 1);
		}

		/* Sort count values */
		uniqueCount = numCounts.size();
		counts = new ArrayList<Integer>(uniqueCount);	// Initialize counts with size of unique ints
		numCounts.forEach((num, count) -> { counts.add(count); });	// Store count values into counts
		Collections.sort(counts);	// Sort counts in ascending order for greedy algorithm

		/* Remove k items */
		int dif = 0;
		for(int i = 0; i < counts.size(); i++) {
			dif = k - counts.get(i);
			if(dif >= 0) {	// There are still elements to be removed
				k = dif;
			}
			else {	// More elements have been removed than allowed, meaning the count left is the solution
				break;
			}

			uniqueCount--;	// Move onto next least common element
		}

		return uniqueCount;
	}   
}

File: 2024/February/2-16 1481 Least Number of Unique Integers After K Removals/LeastUniqueInts_1481.js
var findLeastNumOfUniqueInts = function(arr, k) {
    /* Get count of values in arr */
    let numCounts = new Map();
    for(let i of arr) {
        numCounts.set(i, (numCounts.get(i) || 0) + 1);
    }

    /* Sort count values */
    let counts = Array.from(numCounts.values());
    let uniqueCount = counts.length;
    counts.sort((a, b) => a - b);

    /* Remove k items */
    let dif = 0;
    for(let i = 0; i < counts.length; i++) {
        dif = k - counts[i];
        if(dif >= 0) {
            k = dif;
        }
        else {
            break;
        }

        uniqueCount--;
    }

    return uniqueCount;
};

const nums = [4, 3, 1, 1, 3, 3, 2];
console.log(findLeastNumOfUniqueInts(nums, 3));